src/driver/hermes_mod.h

```diff
@@ -104,6 +104,24 @@ struct hermes_pci_dev {
 	struct xdma_channel xdma_h2c_chnl[XDMA_CHANNEL_NUM_MAX];
 };
 
+static inline int ida_wq_get(struct ida_wq *ida_wq)
+{
+	int id;
+
+	id = ida_alloc_max(&ida_wq->ida, ida_wq->max, GFP_KERNEL);
+	if (id == -ENOSPC)
+		wait_event_interruptible(ida_wq->wq,
+			(id = ida_alloc_max(&ida_wq->ida, ida_wq->max,
+					    GFP_KERNEL)) != -ENOSPC);
```

Comment:
The wait_event_interruptible already contains an if (condition), so some of this is redundant.... 

And you definitely have to handle the return code of wait_event_interruptible()...

But this is all kind of ugly, so roughly something like:

```c
static int __ida_wq_get(struct ida_wq *ida_wq, int *id) 
{
    int ret;

   ret = ida_alloc_max(...);
   if (ret == -ENOSPC)
      return 0;
   *id = ret;
   return 1;
}

static int ida_wq_get(struct ida_wq *ida_wq) 
{
   int id, ret;

   ret = wait_event_interruptible(ida_wq->wq, __ida_wq_get(ida_wq, &id));
   if (ret)
       return ret;
   
  return id; 
}
```